File: aws-lambda-typescript-docker-starter/src/entrypoints/events.ts
import type {
  ALBEvent,
  AmplifyGraphQlResolverEvent,
  APIGatewayAuthorizerEvent,
  APIGatewayProxyEvent,
  APIGatewayProxyEventV2,
  AppSyncResolverEvent,
  AutoScalingScaleInEvent,
  CdkCustomResourceEvent,
  CloudFormationCustomResourceEvent,
  CloudFrontRequestEvent,
  CloudFrontResponseEvent,
  CloudWatchAlarmEvent,
  CloudWatchLogsEvent,
  CodeBuildCloudWatchStateEvent,
  CodeCommitTriggerEvent,
  CodePipelineEvent,
  ConnectContactFlowEvent,
  DynamoDBStreamEvent,
  EventBridgeEvent,
  FirehoseTransformationEvent,
  GuardDutyNotificationEvent,
  IoTCustomAuthorizerEvent,
  IoTEvent,
  KinesisStreamEvent,
  LambdaFunctionURLEvent,
  LexEvent,
  LexV2Event,
  MSKEvent,
  S3BatchEvent,
  S3Event,
  S3NotificationEvent,
  ScheduledEvent,
  SESEvent,
  SelfManagedKafkaEvent,
  SNSEvent,
  SQSEvent,
  SecretsManagerRotationEvent,
  TransferFamilyAuthorizerEvent,
} from 'aws-lambda'

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null
}

function hasString(value: Record<string, unknown>, key: string): boolean {
  return typeof value[key] === 'string'
}

function hasRecordSource(
  event: unknown,
  expectedSource: string,
  sourceKeys: ReadonlyArray<'eventSource' | 'EventSource'> = ['eventSource', 'EventSource'],
): boolean {
  if (!isObject(event) || !Array.isArray(event.Records) || event.Records.length === 0) {
    return false
  }

  const first = event.Records[0]
  if (!isObject(first)) {
    return false
  }

  return sourceKeys.some((key) => first[key] === expectedSource)
}

export function isApiGatewayProxyEvent(event: unknown): event is APIGatewayProxyEvent {
  return isObject(event) && typeof event.httpMethod === 'string'
}

export function isApiGatewayProxyEventV2(event: unknown): event is APIGatewayProxyEventV2 {
  if (!isObject(event) || event.version !== '2.0' || !isObject(event.requestContext)) {
    return false
  }
  const http = event.requestContext.http
  return isObject(http) && typeof http.method === 'string'
}

export function isAlbEvent(event: unknown): event is ALBEvent {
  return isObject(event) && isObject(event.requestContext) && isObject(event.requestContext.elb)
}

export function isHttpEvent(event: unknown): event is APIGatewayProxyEvent | APIGatewayProxyEventV2 | ALBEvent {
  return isAlbEvent(event) || isApiGatewayProxyEvent(event) || isApiGatewayProxyEventV2(event)
}

export function isLambdaFunctionUrlEvent(event: unknown): event is LambdaFunctionURLEvent {
  return isApiGatewayProxyEventV2(event)
}

export function isSqsEvent(event: unknown): event is SQSEvent {
  return hasRecordSource(event, 'aws:sqs', ['eventSource'])
}

export function isSnsEvent(event: unknown): event is SNSEvent {
  return hasRecordSource(event, 'aws:sns', ['EventSource'])
}

export function isS3Event(event: unknown): event is S3Event {
  return hasRecordSource(event, 'aws:s3')
}

export function isSesEvent(event: unknown): event is SESEvent {
  return hasRecordSource(event, 'aws:ses', ['eventSource'])
}

export function isDynamoDbStreamEvent(event: unknown): event is DynamoDBStreamEvent {
  return hasRecordSource(event, 'aws:dynamodb', ['eventSource'])
}

export function isKinesisStreamEvent(event: unknown): event is KinesisStreamEvent {
  return hasRecordSource(event, 'aws:kinesis', ['eventSource'])
}

export function isMskEvent(event: unknown): event is MSKEvent {
  return isObject(event) && event.eventSource === 'aws:kafka' && isObject(event.records)
}

export function isSelfManagedKafkaEvent(event: unknown): event is SelfManagedKafkaEvent {
  return isObject(event) && event.eventSource === 'SelfManagedKafka' && isObject(event.records)
}

export function isCloudWatchLogsEvent(event: unknown): event is CloudWatchLogsEvent {
  return isObject(event) && isObject(event.awslogs) && typeof event.awslogs.data === 'string'
}

export function isEventBridgeEnvelope(event: unknown): event is EventBridgeEvent<string, unknown> {
  return isObject(event) && hasString(event, 'source') && hasString(event, 'detail-type')
}

export function isEventBridgeEvent<TDetailType extends string, TDetail>(
  event: unknown,
  options?: { source?: string; detailType?: TDetailType },
): event is EventBridgeEvent<TDetailType, TDetail> {
  if (!isEventBridgeEnvelope(event)) {
    return false
  }
  if (options?.source && event.source !== options.source) {
    return false
  }
  if (options?.detailType && event['detail-type'] !== options.detailType) {
    return false
  }
  return true
}

export function createEventBridgeEventMatcher<TDetailType extends string, TDetail>(
  options?: { source?: string; detailType?: TDetailType },
): (event: unknown) => event is EventBridgeEvent<TDetailType, TDetail> {
  return (event: unknown): event is EventBridgeEvent<TDetailType, TDetail> =>
    isEventBridgeEvent<TDetailType, TDetail>(event, options)
}

export function isScheduledEvent<TDetail>(event: unknown): event is ScheduledEvent<TDetail> {
  return isEventBridgeEnvelope(event) && event.source === 'aws.events' && event['detail-type'] === 'Scheduled Event'
}

export function isCodeBuildStateEvent(event: unknown): event is CodeBuildCloudWatchStateEvent {
  return (
    isEventBridgeEnvelope(event) &&
    event.source === 'aws.codebuild' &&
    event['detail-type'] === 'CodeBuild Build State Change'
  )
}

export function isCodeCommitEvent(event: unknown): event is CodeCommitTriggerEvent {
  return (
    isObject(event) &&
    Array.isArray(event.Records) &&
    event.Records.length > 0 &&
    isObject(event.Records[0]) &&
    isObject(event.Records[0].codecommit)
  )
}

export function isCodePipelineEvent(event: unknown): event is CodePipelineEvent {
  return isObject(event) && isObject(event['CodePipeline.job'])
}

export function isCloudWatchAlarmEvent(event: unknown): event is CloudWatchAlarmEvent {
  return isObject(event) && hasString(event, 'alarmArn') && isObject(event.alarmData)
}

export function isConnectContactFlowEvent(event: unknown): event is ConnectContactFlowEvent {
  return (
    isObject(event) &&
    event.Name === 'ContactFlowEvent' &&
    isObject(event.Details) &&
    isObject(event.Details.ContactData)
  )
}

export function isCloudFrontRequestEvent(event: unknown): event is CloudFrontRequestEvent {
  return (
    isObject(event) &&
    Array.isArray(event.Records) &&
    event.Records.length > 0 &&
    isObject(event.Records[0]) &&
    isObject(event.Records[0].cf) &&
    isObject(event.Records[0].cf.request) &&
    !isObject(event.Records[0].cf.response)
  )
}

export function isCloudFrontResponseEvent(event: unknown): event is CloudFrontResponseEvent {
  return (
    isObject(event) &&
    Array.isArray(event.Records) &&
    event.Records.length > 0 &&
    isObject(event.Records[0]) &&
    isObject(event.Records[0].cf) &&
    isObject(event.Records[0].cf.response)
  )
}

export function isGuardDutyNotificationEvent(event: unknown): event is GuardDutyNotificationEvent {
  return isEventBridgeEnvelope(event) && event.source === 'aws.guardduty'
}

export function isS3BatchEvent(event: unknown): event is S3BatchEvent {
  return isObject(event) && hasString(event, 'invocationSchemaVersion') && Array.isArray(event.tasks)
}

export function isS3NotificationEvent(event: unknown): event is S3NotificationEvent {
  return isEventBridgeEnvelope(event) && event.source === 'aws.s3'
}

export function isLexEvent(event: unknown): event is LexEvent {
  return isObject(event) && hasString(event, 'messageVersion') && isObject(event.currentIntent) && isObject(event.bot)
}

export function isLexV2Event(event: unknown): event is LexV2Event {
  return isObject(event) && hasString(event, 'messageVersion') && isObject(event.bot) && hasString(event, 'sessionId')
}

export function isAppSyncResolverEvent<TArgs = Record<string, unknown>, TSource = Record<string, unknown> | null>(
  event: unknown,
): event is AppSyncResolverEvent<TArgs, TSource> {
  return isObject(event) && isObject(event.info) && hasString(event.info, 'fieldName') && isObject(event.request)
}

export function isAmplifyResolverEvent<TArgs = Record<string, unknown>, TSource = Record<string, unknown>>(
  event: unknown,
): event is AmplifyGraphQlResolverEvent<TArgs, TSource> {
  return isObject(event) && hasString(event, 'typeName') && hasString(event, 'fieldName') && isObject(event.request)
}

export function isApiGatewayAuthorizerEvent(event: unknown): event is APIGatewayAuthorizerEvent {
  return (
    isObject(event) &&
    (event.type === 'TOKEN' || event.type === 'REQUEST') &&
    (hasString(event, 'methodArn') || hasString(event, 'routeArn'))
  )
}

export function isAutoScalingScaleInEvent(event: unknown): event is AutoScalingScaleInEvent {
  return isObject(event) && hasString(event, 'AutoScalingGroupARN') && Array.isArray(event.CapacityToTerminate)
}

export function isCloudFormationCustomResourceEvent<TProps = Record<string, unknown>, TOldProps = TProps>(
  event: unknown,
): event is CloudFormationCustomResourceEvent<TProps, TOldProps> {
  return (
    isObject(event) &&
    hasString(event, 'RequestType') &&
    hasString(event, 'ResponseURL') &&
    hasString(event, 'StackId') &&
    hasString(event, 'RequestId') &&
    hasString(event, 'LogicalResourceId') &&
    hasString(event, 'ResourceType')
  )
}

export function isCdkCustomResourceEvent<TProps = Record<string, unknown>, TOldProps = TProps>(
  event: unknown,
): event is CdkCustomResourceEvent<TProps, TOldProps> {
  return isCloudFormationCustomResourceEvent<TProps, TOldProps>(event)
}

export function isIoTEvent<TPayload = never>(event: unknown): event is IoTEvent<TPayload> {
  return typeof event === 'string' || typeof event === 'number'
}

export function isIoTCustomAuthorizerEvent(event: unknown): event is IoTCustomAuthorizerEvent {
  return isObject(event) && isObject(event.protocolData) && isObject(event.connectionMetadata)
}

export function isFirehoseTransformationEvent(event: unknown): event is FirehoseTransformationEvent {
  return isObject(event) && hasString(event, 'invocationId') && Array.isArray(event.records)
}

export function isSecretsManagerRotationEvent(event: unknown): event is SecretsManagerRotationEvent {
  return (
    isObject(event) &&
    hasString(event, 'Step') &&
    hasString(event, 'SecretId') &&
    hasString(event, 'ClientRequestToken')
  )
}

export function isTransferFamilyAuthorizerEvent(event: unknown): event is TransferFamilyAuthorizerEvent {
  return isObject(event) && hasString(event, 'username') && hasString(event, 'serverId') && hasString(event, 'sourceIp')
}

File: aws-lambda-golang-docker-starter/internal/lambdaapp/events.go
package lambdaapp

import (
	"encoding/json"
	"strings"

	awsevents "github.com/aws/aws-lambda-go/events"
)

type EventType string

const (
	EventTypeHTTPV2                 EventType = "http-v2"
	EventTypeLambdaFunctionURL      EventType = "lambda-function-url"
	EventTypeALB                    EventType = "alb"
	EventTypeHTTPV1                 EventType = "http-v1"
	EventTypeSQS                    EventType = "sqs"
	EventTypeSNS                    EventType = "sns"
	EventTypeS3                     EventType = "s3"
	EventTypeSES                    EventType = "ses"
	EventTypeDynamoDBStream         EventType = "dynamodb-stream"
	EventTypeKinesisStream          EventType = "kinesis-stream"
	EventTypeCloudWatchLogs         EventType = "cloudwatch-logs"
	EventTypeScheduled              EventType = "scheduled"
	EventTypeCodeBuildState         EventType = "codebuild-state"
	EventTypeCodeCommit             EventType = "codecommit"
	EventTypeEventBridge            EventType = "eventbridge"
	EventTypeCodePipeline           EventType = "codepipeline"
	EventTypeCloudWatchAlarm        EventType = "cloudwatch-alarm"
	EventTypeConnectContactFlow     EventType = "connect-contact-flow"
	EventTypeCloudFrontRequest      EventType = "cloudfront-request"
	EventTypeCloudFrontResponse     EventType = "cloudfront-response"
	EventTypeGuardDutyNotification  EventType = "guardduty-notification"
	EventTypeS3Batch                EventType = "s3-batch"
	EventTypeS3Notification         EventType = "s3-notification"
	EventTypeLex                    EventType = "lex"
	EventTypeLexV2                  EventType = "lex-v2"
	EventTypeAppSyncResolver        EventType = "appsync-resolver"
	EventTypeAmplifyResolver        EventType = "amplify-resolver"
	EventTypeAPIGatewayAuthorizer   EventType = "api-gateway-authorizer"
	EventTypeAutoScalingScaleIn     EventType = "autoscaling-scale-in"
	EventTypeCloudFormationCustom   EventType = "cloudformation-custom-resource"
	EventTypeCdkCustomResource      EventType = "cdk-custom-resource"
	EventTypeIoT                    EventType = "iot"
	EventTypeIoTCustomAuthorizer    EventType = "iot-custom-authorizer"
	EventTypeFirehoseTransformation EventType = "firehose-transformation"
	EventTypeSecretsManagerRotation EventType = "secrets-manager-rotation"
	EventTypeTransferFamilyAuth     EventType = "transfer-family-authorizer"
	EventTypeMSK                    EventType = "msk"
	EventTypeSelfManagedKafka       EventType = "self-managed-kafka"
	EventTypeUnknown                EventType = "unknown"
)

func detectEventType(raw json.RawMessage) (EventType, error) {
	// Parse once up front so malformed JSON always returns a hard error.
	var payload any
	if err := json.Unmarshal(raw, &payload); err != nil {
		return EventTypeUnknown, err
	}

	// IoT events can be scalar payloads (string/number).
	switch payload.(type) {
	case string, float64:
		return EventTypeIoT, nil
	}

	envelope, ok := payload.(map[string]any)
	if !ok {
		return EventTypeUnknown, nil
	}

	switch {
	case matches(raw, func(event awsevents.APIGatewayV2HTTPRequest) bool {
		return event.Version == "2.0" && strings.TrimSpace(event.RequestContext.HTTP.Method) != ""
	}):
		if domainName := strings.TrimSpace(toString(envelope["requestContext"], "domainName")); strings.Contains(domainName, "lambda-url") {
			return EventTypeLambdaFunctionURL, nil
		}
		return EventTypeHTTPV2, nil
	case nestedObject(envelope, "requestContext", "elb") != nil:
		return EventTypeALB, nil
	case matches(raw, func(event awsevents.APIGatewayProxyRequest) bool {
		return strings.TrimSpace(event.HTTPMethod) != ""
	}):
		return EventTypeHTTPV1, nil
	case matches(raw, func(event awsevents.SQSEvent) bool {
		return len(event.Records) > 0 && event.Records[0].EventSource == "aws:sqs"
	}):
		return EventTypeSQS, nil
	case matches(raw, func(event awsevents.SNSEvent) bool {
		return len(event.Records) > 0 && event.Records[0].EventSource == "aws:sns"
	}):
		return EventTypeSNS, nil
	case matches(raw, func(event awsevents.S3Event) bool {
		return len(event.Records) > 0 && event.Records[0].EventSource == "aws:s3"
	}):
		return EventTypeS3, nil
	case matches(raw, func(event awsevents.SimpleEmailEvent) bool {
		return len(event.Records) > 0 && event.Records[0].EventSource == "aws:ses"
	}):
		return EventTypeSES, nil
	case matches(raw, func(event awsevents.DynamoDBEvent) bool {
		return len(event.Records) > 0 && event.Records[0].EventSource == "aws:dynamodb"
	}):
		return EventTypeDynamoDBStream, nil
	case matches(raw, func(event awsevents.KinesisEvent) bool {
		return len(event.Records) > 0 && event.Records[0].EventSource == "aws:kinesis"
	}):
		return EventTypeKinesisStream, nil
	case matches(raw, func(event awsevents.CloudwatchLogsEvent) bool {
		return strings.TrimSpace(event.AWSLogs.Data) != ""
	}):
		return EventTypeCloudWatchLogs, nil
	case matches(raw, func(event awsevents.CloudWatchEvent) bool {
		return event.Source == "aws.events" && event.DetailType == "Scheduled Event"
	}):
		return EventTypeScheduled, nil
	case matches(raw, func(event awsevents.CodeBuildEvent) bool {
		return event.Source == "aws.codebuild" && event.DetailType == "CodeBuild Build State Change"
	}):
		return EventTypeCodeBuildState, nil
	case hasCodeCommitRecord(envelope):
		return EventTypeCodeCommit, nil
	case hasString(envelope, "alarmArn") && isObject(envelope["alarmData"]):
		return EventTypeCloudWatchAlarm, nil
	case envelope["Name"] == "ContactFlowEvent" && isObject(nestedObject(envelope, "Details", "ContactData")):
		return EventTypeConnectContactFlow, nil
	case isCloudFrontRequestEvent(envelope):
		return EventTypeCloudFrontRequest, nil
	case isCloudFrontResponseEvent(envelope):
		return EventTypeCloudFrontResponse, nil
	case matches(raw, func(event awsevents.CloudWatchEvent) bool {
		return event.Source == "aws.guardduty"
	}):
		return EventTypeGuardDutyNotification, nil
	case hasString(envelope, "invocationSchemaVersion") && isArray(envelope["tasks"]):
		return EventTypeS3Batch, nil
	case matches(raw, func(event awsevents.CloudWatchEvent) bool {
		return event.Source == "aws.s3" && strings.TrimSpace(event.DetailType) != ""
	}):
		return EventTypeS3Notification, nil
	case hasString(envelope, "messageVersion") && isObject(envelope["currentIntent"]) && isObject(envelope["bot"]):
		return EventTypeLex, nil
	case hasString(envelope, "messageVersion") && isObject(envelope["bot"]) && hasString(envelope, "sessionId"):
		return EventTypeLexV2, nil
	case hasString(nestedObject(envelope, "info"), "fieldName") && isObject(envelope["request"]):
		return EventTypeAppSyncResolver, nil
	case hasString(envelope, "typeName") && hasString(envelope, "fieldName") && isObject(envelope["request"]):
		return EventTypeAmplifyResolver, nil
	case isAPIGatewayAuthorizerEvent(envelope):
		return EventTypeAPIGatewayAuthorizer, nil
	case hasString(envelope, "AutoScalingGroupARN") && isArray(envelope["CapacityToTerminate"]):
		return EventTypeAutoScalingScaleIn, nil
	case isCloudFormationCustomResourceEvent(envelope):
		// CDK custom resource event shares the same envelope shape.
		return EventTypeCloudFormationCustom, nil
	case isObject(envelope["protocolData"]) && isObject(envelope["connectionMetadata"]):
		return EventTypeIoTCustomAuthorizer, nil
	case matches(raw, func(event awsevents.CloudWatchEvent) bool {
		return strings.TrimSpace(event.Source) != "" && strings.TrimSpace(event.DetailType) != ""
	}):
		return EventTypeEventBridge, nil
	case matches(raw, func(event awsevents.CodePipelineEvent) bool {
		return strings.TrimSpace(event.CodePipelineJob.ID) != ""
	}):
		return EventTypeCodePipeline, nil
	case matches(raw, func(event awsevents.KinesisFirehoseEvent) bool {
		return strings.TrimSpace(event.InvocationID) != "" && len(event.Records) > 0
	}):
		return EventTypeFirehoseTransformation, nil
	case matches(raw, func(event awsevents.SecretsManagerSecretRotationEvent) bool {
		return strings.TrimSpace(event.Step) != "" &&
			strings.TrimSpace(event.SecretID) != "" &&
			strings.TrimSpace(event.ClientRequestToken) != ""
	}):
		return EventTypeSecretsManagerRotation, nil
	case hasString(envelope, "username") && hasString(envelope, "serverId") && hasString(envelope, "sourceIp"):
		return EventTypeTransferFamilyAuth, nil
	case matches(raw, func(event awsevents.KafkaEvent) bool {
		return event.EventSource == "aws:kafka" && len(event.Records) > 0
	}):
		return EventTypeMSK, nil
	case matches(raw, func(event awsevents.KafkaEvent) bool {
		return event.EventSource == "SelfManagedKafka" && len(event.Records) > 0
	}):
		return EventTypeSelfManagedKafka, nil
	}

	return EventTypeUnknown, nil
}

func matches[T any](raw json.RawMessage, matcher func(event T) bool) bool {
	var event T
	if err := json.Unmarshal(raw, &event); err != nil {
		return false
	}
	return matcher(event)
}

func isObject(v any) bool {
	_, ok := v.(map[string]any)
	return ok
}

func isArray(v any) bool {
	_, ok := v.([]any)
	return ok
}

func nestedObject(root map[string]any, keys ...string) map[string]any {
	current := root
	for _, key := range keys {
		next, ok := current[key].(map[string]any)
		if !ok {
			return nil
		}
		current = next
	}
	return current
}

func toString(v any, keys ...string) string {
	obj, ok := v.(map[string]any)
	if !ok {
		return ""
	}
	current := obj
	for i, key := range keys {
		value, exists := current[key]
		if !exists {
			return ""
		}
		if i == len(keys)-1 {
			s, _ := value.(string)
			return s
		}
		next, ok := value.(map[string]any)
		if !ok {
			return ""
		}
		current = next
	}
	return ""
}

func hasString(root map[string]any, key string) bool {
	_, ok := root[key].(string)
	return ok
}

func hasCodeCommitRecord(envelope map[string]any) bool {
	records, ok := envelope["Records"].([]any)
	if !ok || len(records) == 0 {
		return false
	}
	first, ok := records[0].(map[string]any)
	if !ok {
		return false
	}
	return isObject(first["codecommit"])
}

func isCloudFrontRequestEvent(envelope map[string]any) bool {
	records, ok := envelope["Records"].([]any)
	if !ok || len(records) == 0 {
		return false
	}
	first, ok := records[0].(map[string]any)
	if !ok {
		return false
	}
	cf, ok := first["cf"].(map[string]any)
	if !ok {
		return false
	}
	_, hasReq := cf["request"].(map[string]any)
	_, hasResp := cf["response"].(map[string]any)
	return hasReq && !hasResp
}

func isCloudFrontResponseEvent(envelope map[string]any) bool {
	records, ok := envelope["Records"].([]any)
	if !ok || len(records) == 0 {
		return false
	}
	first, ok := records[0].(map[string]any)
	if !ok {
		return false
	}
	cf, ok := first["cf"].(map[string]any)
	if !ok {
		return false
	}
	_, hasResp := cf["response"].(map[string]any)
	return hasResp
}

func isAPIGatewayAuthorizerEvent(envelope map[string]any) bool {
	eventType, ok := envelope["type"].(string)
	if !ok || (eventType != "TOKEN" && eventType != "REQUEST") {
		return false
	}
	return hasString(envelope, "methodArn") || hasString(envelope, "routeArn")
}

func isCloudFormationCustomResourceEvent(envelope map[string]any) bool {
	required := []string{
		"RequestType",
		"ResponseURL",
		"StackId",
		"RequestId",
		"LogicalResourceId",
		"ResourceType",
	}
	for _, key := range required {
		if !hasString(envelope, key) {
			return false
		}
	}
	return true
}
